File: src/simulation/pools.ts
import {
  bullets,
  chests,
  counters,
  enemies,
  experienceOrbs,
  floaters,
  particles,
  powerupOrbs,
} from "../state";
import type {
  Bullet,
  ChestEntity,
  EnemyEntity,
  EnemyKind,
  EnemyType,
  ExperienceOrb,
  Floater,
  Particle,
  PowerupKind,
  PowerupOrb,
} from "../types";
import { swapRemove } from "../utils";

const enemyPool: EnemyEntity[] = [];
const bulletPool: Bullet[] = [];
const chestPool: ChestEntity[] = [];
const experiencePool: ExperienceOrb[] = [];
const powerupPool: PowerupOrb[] = [];
const particlePool: Particle[] = [];
const floaterPool: Floater[] = [];

export function clearEntityPools(): void {
  enemyPool.length = 0;
  bulletPool.length = 0;
  chestPool.length = 0;
  experiencePool.length = 0;
  powerupPool.length = 0;
  particlePool.length = 0;
  floaterPool.length = 0;
}

export function resetEntityCounters(): void {
  counters.nextEnemyId = 1;
  counters.nextBulletId = 1;
  counters.nextExperienceId = 1;
  counters.nextPowerupId = 1;
  counters.nextChestId = 1;
  counters.nextParticleId = 1;
  counters.nextFloaterId = 1;
}

export function acquireEnemy(type: EnemyType, kind: EnemyKind): EnemyEntity {
  const enemy =
    enemyPool.pop() ??
    ({
      id: 0,
      kind,
      score: 0,
      radius: 0,
      hp: 0,
      maxHp: 0,
      speed: 0,
      damage: 0,
      color: "#ffffff",
      accent: "#ffffff",
      sides: 3,
      x: 0,
      y: 0,
      age: 0,
      seed: 0,
      wobble: 0,
      wobbleRate: 0,
      hit: 0,
    } satisfies EnemyEntity);
  enemy.id = counters.nextEnemyId;
  counters.nextEnemyId += 1;
  enemy.kind = kind;
  enemy.score = type.score;
  enemy.radius = type.radius;
  enemy.damage = type.damage;
  enemy.color = type.color;
  enemy.accent = type.accent;
  enemy.sides = type.sides;
  enemy.role = "normal";
  enemy.bossVariant = undefined;
  enemy.contactTimer = 0;
  enemy.contactCooldown = undefined;
  enemies.push(enemy);
  return enemy;
}

export function releaseEnemy(index: number): EnemyEntity {
  const enemy = swapRemove(enemies, index);
  enemyPool.push(enemy);
  return enemy;
}

export function acquireBullet(): Bullet {
  const bullet =
    bulletPool.pop() ??
    ({
      id: 0,
      x: 0,
      y: 0,
      vx: 0,
      vy: 0,
      radius: 0,
      damage: 0,
      pierce: 0,
      life: 0,
      color: "#ffffff",
      trail: 0,
      hitIds: new Set<number>(),
      source: "player",
      chainRemaining: 0,
    } satisfies Bullet);
  bullet.id = counters.nextBulletId;
  counters.nextBulletId += 1;
  bullet.hitIds.clear();
  bullet.source = "player";
  bullet.chainRemaining = 0;
  bullets.push(bullet);
  return bullet;
}

export function releaseBullet(index: number): Bullet {
  const bullet = swapRemove(bullets, index);
  bullet.hitIds.clear();
  bulletPool.push(bullet);
  return bullet;
}

export function acquireChest(): ChestEntity {
  const chest =
    chestPool.pop() ??
    ({
      id: 0,
      x: 0,
      y: 0,
      vx: 0,
      vy: 0,
      radius: 0,
      age: 0,
    } satisfies ChestEntity);
  chest.id = counters.nextChestId;
  counters.nextChestId += 1;
  chests.push(chest);
  return chest;
}

export function releaseChest(index: number): ChestEntity {
  const chest = swapRemove(chests, index);
  chestPool.push(chest);
  return chest;
}

export function acquireExperienceOrb(): ExperienceOrb {
  const orb =
    experiencePool.pop() ??
    ({
      id: 0,
      x: 0,
      y: 0,
      vx: 0,
      vy: 0,
      radius: 0,
      value: 0,
      age: 0,
      magnetized: false,
    } satisfies ExperienceOrb);
  orb.id = counters.nextExperienceId;
  counters.nextExperienceId += 1;
  experienceOrbs.push(orb);
  return orb;
}

export function releaseExperienceOrb(index: number): ExperienceOrb {
  const orb = swapRemove(experienceOrbs, index);
  experiencePool.push(orb);
  return orb;
}

export function acquirePowerupOrb(kind: PowerupKind): PowerupOrb {
  const orb =
    powerupPool.pop() ??
    ({
      id: 0,
      x: 0,
      y: 0,
      vx: 0,
      vy: 0,
      radius: 0,
      kind,
      age: 0,
      life: 0,
    } satisfies PowerupOrb);
  orb.id = counters.nextPowerupId;
  counters.nextPowerupId += 1;
  orb.kind = kind;
  powerupOrbs.push(orb);
  return orb;
}

export function releasePowerupOrb(index: number): PowerupOrb {
  const orb = swapRemove(powerupOrbs, index);
  powerupPool.push(orb);
  return orb;
}

export function acquireParticle(): Particle {
  const particle =
    particlePool.pop() ??
    ({
      id: 0,
      x: 0,
      y: 0,
      vx: 0,
      vy: 0,
      size: 0,
      color: "#ffffff",
      life: 0,
      maxLife: 0,
      behind: false,
    } satisfies Particle);
  particle.id = counters.nextParticleId;
  counters.nextParticleId += 1;
  particles.push(particle);
  return particle;
}

export function releaseParticle(index: number): Particle {
  const particle = swapRemove(particles, index);
  particlePool.push(particle);
  return particle;
}

export function acquireFloater(): Floater {
  const floater =
    floaterPool.pop() ??
    ({
      id: 0,
      x: 0,
      y: 0,
      text: "",
      color: "#ffffff",
      damageText: false,
      life: 0,
      maxLife: 0,
    } satisfies Floater);
  floater.id = counters.nextFloaterId;
  counters.nextFloaterId += 1;
  floaters.push(floater);
  return floater;
}

export function releaseFloater(index: number): Floater {
  const floater = swapRemove(floaters, index);
  floaterPool.push(floater);
  return floater;
}

File: sim/crates/voidline-sim/src/pools.rs
//! Entity pools for the Rust runtime engine. Pools recycle entities to avoid
//! allocations while keeping release behavior deterministic via swap_remove.

use crate::entities::{
    Bullet, BulletSource, ChestEntity, Enemy, EnemyKind, ExperienceOrb, PowerupKind, PowerupOrb,
};
use crate::state::EntityCounters;

#[derive(Debug, Default)]
pub struct EntityPools {
    pub enemy_pool: Vec<Enemy>,
    pub bullet_pool: Vec<Bullet>,
    pub chest_pool: Vec<ChestEntity>,
    pub experience_pool: Vec<ExperienceOrb>,
    pub powerup_pool: Vec<PowerupOrb>,
}

impl EntityPools {
    pub fn clear(&mut self) {
        self.enemy_pool.clear();
        self.bullet_pool.clear();
        self.chest_pool.clear();
        self.experience_pool.clear();
        self.powerup_pool.clear();
    }
}

pub fn acquire_enemy(
    pools: &mut EntityPools,
    counters: &mut EntityCounters,
    enemies: &mut Vec<Enemy>,
    kind: EnemyKind,
    score: f64,
    radius: f64,
    damage: f64,
    sides: u32,
) -> usize {
    let mut enemy = pools.enemy_pool.pop().unwrap_or(Enemy {
        id: 0,
        kind,
        score: 0.0,
        radius: 0.0,
        hp: 0.0,
        max_hp: 0.0,
        speed: 0.0,
        damage: 0.0,
        sides: 3,
        x: 0.0,
        y: 0.0,
        age: 0.0,
        seed: 0.0,
        wobble: 0.0,
        wobble_rate: 0.0,
        hit: 0.0,
        role: crate::entities::EnemyRole::Normal,
        contact_timer: 0.0,
        contact_cooldown: 0.0,
    });
    enemy.id = counters.next_enemy_id;
    counters.next_enemy_id += 1;
    enemy.kind = kind;
    enemy.score = score;
    enemy.radius = radius;
    enemy.damage = damage;
    enemy.sides = sides;
    enemy.role = crate::entities::EnemyRole::Normal;
    enemy.contact_timer = 0.0;
    enemy.contact_cooldown = 0.0;
    enemies.push(enemy);
    enemies.len() - 1
}

/// Mirrors `swapRemove` semantics: O(1) by swapping with last.
pub fn release_enemy(pools: &mut EntityPools, enemies: &mut Vec<Enemy>, index: usize) -> Enemy {
    let enemy = enemies.swap_remove(index);
    let mut returned = enemy.clone();
    pools.enemy_pool.push(enemy);
    returned.role = crate::entities::EnemyRole::Normal;
    returned
}

pub fn acquire_bullet(
    pools: &mut EntityPools,
    counters: &mut EntityCounters,
    bullets: &mut Vec<Bullet>,
) -> usize {
    let mut bullet = pools.bullet_pool.pop().unwrap_or(Bullet {
        id: 0,
        x: 0.0,
        y: 0.0,
        vx: 0.0,
        vy: 0.0,
        radius: 0.0,
        damage: 0.0,
        pierce: 0,
        life: 0.0,
        trail: 0.0,
        hit_ids: Vec::new(),
        source: BulletSource::Player,
        chain_remaining: 0,
    });
    bullet.id = counters.next_bullet_id;
    counters.next_bullet_id += 1;
    bullet.hit_ids.clear();
    bullet.source = BulletSource::Player;
    bullet.chain_remaining = 0;
    bullets.push(bullet);
    bullets.len() - 1
}

pub fn release_bullet(pools: &mut EntityPools, bullets: &mut Vec<Bullet>, index: usize) {
    let mut bullet = bullets.swap_remove(index);
    bullet.hit_ids.clear();
    pools.bullet_pool.push(bullet);
}

pub fn acquire_chest(
    pools: &mut EntityPools,
    counters: &mut EntityCounters,
    chests: &mut Vec<ChestEntity>,
) -> usize {
    let mut chest = pools.chest_pool.pop().unwrap_or(ChestEntity {
        id: 0,
        x: 0.0,
        y: 0.0,
        vx: 0.0,
        vy: 0.0,
        radius: 0.0,
        age: 0.0,
    });
    chest.id = counters.next_chest_id;
    counters.next_chest_id += 1;
    chests.push(chest);
    chests.len() - 1
}

pub fn release_chest(pools: &mut EntityPools, chests: &mut Vec<ChestEntity>, index: usize) {
    let chest = chests.swap_remove(index);
    pools.chest_pool.push(chest);
}

pub fn acquire_experience_orb(
    pools: &mut EntityPools,
    counters: &mut EntityCounters,
    orbs: &mut Vec<ExperienceOrb>,
) -> usize {
    let mut orb = pools.experience_pool.pop().unwrap_or(ExperienceOrb {
        id: 0,
        x: 0.0,
        y: 0.0,
        vx: 0.0,
        vy: 0.0,
        radius: 0.0,
        value: 0.0,
        age: 0.0,
        magnetized: false,
    });
    orb.id = counters.next_experience_id;
    counters.next_experience_id += 1;
    orbs.push(orb);
    orbs.len() - 1
}

pub fn release_experience_orb(
    pools: &mut EntityPools,
    orbs: &mut Vec<ExperienceOrb>,
    index: usize,
) {
    let orb = orbs.swap_remove(index);
    pools.experience_pool.push(orb);
}

pub fn acquire_powerup_orb(
    pools: &mut EntityPools,
    counters: &mut EntityCounters,
    orbs: &mut Vec<PowerupOrb>,
    kind: PowerupKind,
) -> usize {
    let mut orb = pools.powerup_pool.pop().unwrap_or(PowerupOrb {
        id: 0,
        x: 0.0,
        y: 0.0,
        vx: 0.0,
        vy: 0.0,
        radius: 0.0,
        kind,
        age: 0.0,
        life: 0.0,
    });
    orb.id = counters.next_powerup_id;
    counters.next_powerup_id += 1;
    orb.kind = kind;
    orbs.push(orb);
    orbs.len() - 1
}

pub fn release_powerup_orb(pools: &mut EntityPools, orbs: &mut Vec<PowerupOrb>, index: usize) {
    let orb = orbs.swap_remove(index);
    pools.powerup_pool.push(orb);
}
